content/blog/dynamic-providers/index.md

```diff
@@ -0,0 +1,114 @@
+---
+title: "Dynamic Providers"
+date: 2019-12-20
+meta_desc: "Pulumi dynamic providers manage resources that do not have prebuilt providers such as AWS, Azure, and GCP"
+meta_image: meta.png
+authors:
+    - praneet-loke
+tags:
+    - providers
+    - resources
+---
+
+Pulumi has several resource providers (over 45!) that allow you to interact with your favorite cloud provider or resource using a programming language. But what if there isn’t a provider available for what you want to accomplish? That’s where dynamic providers can help you. They can be an extremely important part of your infrastructure app.
+
+<!--more-->
+
+A provider manages the CRUD lifecycle of a resource against some underlying service. The underlying service can be one of the public clouds or an on-premise service. For example, AWS, Azure, GCP, and Kubernetes are all resource providers, which can create/update/destroy resources in the respective clouds.
+
+## _Dynamic_ Resource Provider
+
+> **Note**: Dynamic providers are currently only supported in Javascript, Typescript, and Python.
+
+Pulumi provides appropriate lifecycle callbacks through the `pulumi.dynamic.ResourceProvider` abstract class. While Pulumi does not know what sort of resources a dynamic provider will create, the lifecycle hooks must return data in specific formats, which Pulumi uses as hints to call the appropriate lifecycle hooks.
+
+For example, if the user changes an input property to your dynamic resource, the `diff` hook will be called with both the old (if applicable) and new inputs. You can then compare to see which input properties have changed and give hints to Pulumi as to whether `delete` should be called or whether to call the `update` hook.
+
+> The `check` lifecycle hook will be called before any other hook to let you validate the inputs and populate defaults, if necessary.
+
+When a dynamic resource instance updates, the `delete` hook will be called to allow you to perform any necessary cleanups before the dynamic resource is removed from your stack’s state.
+
+### Provider Serialization
+
+It is important to note that once an instance of the dynamic resource is created, Pulumi serializes the provider and stores it in your [state](https://www.pulumi.com/docs/intro/concepts/state/#state) file. When you change the source code of your dynamic resource, the serialized version is updated.
+
+## Use Cases
+
+Dynamic providers can be used for almost any task. Here are some use cases:
```

Comment:
I would discuss these possibilities as we go through examples, and then generalized in the conclusion at the end of the article